### ISSUE TITLE
upstream module with balancer_by_lua and keepalive_requests in OpenResty 1.15.8.1 not working

### ISSUE THREAD

**Issue by bmgoau** [openresty/lua-nginx-module#1532]
Hello Team,

We've always seen that 1% of upstream requests result in a new TCP connection, which makes sense, since the default for keepalive_requests has always been 100.
http://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive_requests

Today we tried setting keepalive_requests 100000; expecting that (given our traffic load per worker is sufficiently high and busy) new connections would shrink below 1%.

However, even after deploying the new configuration, with no errors, we still see 1% of requests result in a new upstream connection, and upstream keepalives ending after 100 requests.

We have upstream configuration like this:

upstream backend {
    server 0.0.0.1; # placeholder
    balancer_by_lua_file /etc/nginx/lua/upstream.lua;
    keepalive 10000; # default no keepalives, per worker
    keepalive_requests 100000; # default is 100
    keepalive_timeout 59; # default is 60s, must be lower than upstream
}

We suspect that balancer_by_lua_file is not respecting keepalive_requests custom value.

Can anyone shed any light on this?

**Comment by agentzh** [comment]
@bmgoau Your backend server may also have a limit on the keepalive request count in a single connection. Check your backend server's configuration and/or the TCP packets on the link.

**Comment by bmgoau** [comment]
Hello @agentzh  thanks for your quick reply and amazing work here!

In this case our upstream is an AWS ALB with three IPs (confirmed using the all. prefix on the fqdn), and we've confirmed with AWS that although the ALB has a keepalive idle timeout of 60 seconds (we've tried raising this to no effect), there is no cap on the number of requests the ALB will accept over a connection. It is effectively unlimited.

Per our configuration we have an ngingx keepalive timeout of 59 seconds
This particular openresty instance sends around 1055 requests per minute to this upstream from 8 workers, so 131.875 requests per minute per worker.
These are balanced across the three upstream IPs so there should be around 43 RPM per worker per upstream IP - enough to keep a keepalive sessions from both the nginx and ALB perspective open continuously.

I think the tell-tale fact here is that exactly 1% of requests result in a new TCP connection with the ALB - which would seem to imply that keepalive_requests 100000; is not having any effect.
